File: backend/app/schemas/battery.py
"""Pydantic mirror of shared/schemas/battery_metadata.schema.json.

Keep in sync with that file manually - see shared/README.md.
"""

from enum import StrEnum

from pydantic import BaseModel, Field


class Chemistry(StrEnum):
    NMC = "NMC"
    NCA = "NCA"
    LFP = "LFP"
    LMO = "LMO"
    LTO = "LTO"
    OTHER = "Other"


class FormFactor(StrEnum):
    CYLINDRICAL = "cylindrical"
    POUCH = "pouch"
    PRISMATIC = "prismatic"


class TemperatureRange(BaseModel):
    min: float
    max: float


class BatteryMetadata(BaseModel):
    battery_id: str = Field(..., min_length=1)
    manufacturer: str | None = None
    model_name: str | None = None
    nominal_capacity_kwh: float = Field(
        ...,
        gt=0,
        description=(
            "Rated pack energy capacity in kWh. Intentionally unconstrained - "
            "any pack size is accepted, not just the reference sizes used "
            "during model development."
        ),
    )
    nominal_voltage_v: float = Field(..., gt=0)
    chemistry: Chemistry
    pack_configuration: str | None = None
    cell_count: int | None = Field(default=None, ge=1)
    form_factor: FormFactor | None = None
    manufacture_year: int | None = Field(default=None, ge=1990)
    cycle_count_at_upload: int | None = Field(default=None, ge=0)
    operating_temperature_range_c: TemperatureRange | None = None
    notes: str | None = None

File: frontend/src/types/battery.ts
/**
 * TypeScript mirror of shared/schemas/battery_metadata.schema.json.
 * Keep in sync with that file and backend/app/schemas/battery.py manually -
 * see shared/README.md.
 */

export type Chemistry = 'NMC' | 'NCA' | 'LFP' | 'LMO' | 'LTO' | 'Other'

export type FormFactor = 'cylindrical' | 'pouch' | 'prismatic'

export interface TemperatureRange {
  min: number
  max: number
}

export interface BatteryMetadata {
  battery_id: string
  manufacturer?: string | null
  model_name?: string | null
  /**
   * Rated pack energy capacity in kWh. Intentionally unconstrained - any
   * pack size is accepted, not just the reference sizes used during model
   * development.
   */
  nominal_capacity_kwh: number
  nominal_voltage_v: number
  chemistry: Chemistry
  pack_configuration?: string | null
  cell_count?: number | null
  form_factor?: FormFactor | null
  manufacture_year?: number | null
  cycle_count_at_upload?: number | null
  operating_temperature_range_c?: TemperatureRange | null
  notes?: string | null
}
